tpke/src/combine.rs

```diff
@@ -2,71 +2,100 @@
 #![allow(dead_code)]
 use crate::*;
 use ark_ec::ProjectiveCurve;
+use itertools::zip_eq;
 
 pub fn prepare_combine<E: PairingEngine>(
     public_decryption_contexts: &[PublicDecryptionContext<E>],
     shares: &[DecryptionShare<E>],
 ) -> Vec<E::G2Prepared> {
-    let mut domain = vec![];
+    let mut domain = vec![]; // omega_i, vector of domain points
     let mut n_0 = E::Fr::one();
     for d_i in shares.iter() {
-        domain.extend(
-            public_decryption_contexts[d_i.decrypter_index]
-                .domain
-                .iter(),
+        // There's just one domain point per participant, TODO: Refactor underlying data structures
+        assert_eq!(
+            public_decryption_contexts[d_i.decrypter_index].domain.len(),
+            1
         );
-        n_0 *= public_decryption_contexts[d_i.decrypter_index].lagrange_n_0;
+        domain.push(public_decryption_contexts[d_i.decrypter_index].domain[0]);
+        n_0 *= public_decryption_contexts[d_i.decrypter_index].lagrange_n_0; // n_0_i = 1 * t^1 * t^2 ...
     }
     let s = SubproductDomain::<E::Fr>::new(domain);
-    let mut lagrange = s.inverse_lagrange_coefficients();
-    ark_ff::batch_inversion_and_mul(&mut lagrange, &n_0);
-    let mut start = 0usize;
-    shares
-        .iter()
-        .map(|d_i| {
+    let mut lagrange = s.inverse_lagrange_coefficients(); // 1/L_i
```

Comment:
why do we need the inverse coefficients here, but not in `prepare_combine_simple`? I'm guessing it comes from one of the papers that explains #5

Reply:
This inversion is specific to the fast decryption variant since it uses [Subproduct Domain technique](https://nikkolasg.github.io/ferveo/lagrange.html#lagrange-interpolation):

"where $\frac{1}{λ_i}$ is the inverse Lagrange coefficient as computed using the Subproduct Domain technique"

Reply:
ah thank you! I feel like an undergrad again, this is great fun!